Adds foreman fee counter for Blast Furnace under level 60 smithing

bwhite94 opened the pull request:
For players that are utilizing blast furnace and require to pay the
foreman fee every 10 minutes, this will assist in tracking the time
remaining before the next fee payment is required. This defaults to
disabled, but is bundled with the Blast Furnace plugin config.

Resolves #8769

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -50,12 +59,17 @@
 {
 	private static final int BAR_DISPENSER = NULL_9092;
 
+	private static final Pattern FOREMAN_PERMISSION_PATTERN = Pattern.compile("Okay, you can use the furnace for ten minutes. Remember, you only need half as much coal as with a regular furnace.");
```
Why are you using a `Pattern` for a static string?

bwhite94 replied:
Resolved in d642d6b

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -25,21 +25,30 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static net.runelite.api.NullObjectID.NULL_9092;
```
Why are these object IDs imported?

bwhite94 replied:
Ah, it looks.like this was just shifted by IntelliJ when importing other classes. This is used in other portions of the plug in that are pre-existing.

Nightfirecat replied:
Well, keep in mind that imports should all be in one sorted block without extra newlines separating them.

bwhite94 replied:
Resolved in d642d6b

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -135,4 +158,27 @@ public void onGameStateChanged(GameStateChanged event)
+
+			if (mPermission.find())
+			{
+				foremanCounter = 100 * 10;
```
Define the duration as a final static variable, and use a `Duration` instead of measuring in ticks.

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -135,4 +158,27 @@ public void onGameStateChanged(GameStateChanged event)
+
+		if (foremanCounter != null && foremanCounter > 0)
+		{
+			foremanCounter--;
```
Rather than measuring in ticks, save the time at which the foreman permission text is last found, then compute the time remaining in the overlay's render function. The time left to display should be the static duration (10 minutes) minus the difference between that saved timestamp and the current time.

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceForemanOverlay.java:
```diff
@@ -0,0 +1,73 @@
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (config.showForemanCounter())
```
Use a guard clause instead:

```java
if (!config.showForemanCounter())
{
	return null;
}

// ...rest of render code goes here
```

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceForemanOverlay.java:
```diff
@@ -0,0 +1,73 @@
+public class BlastFurnaceForemanOverlay extends Overlay
+{
+
+	private final Client client;
```
Why are you injecting a `Client` here? It doesn't seem to be used.

bwhite94 replied:
Now utilized for smithing skill lookup. Left in-tact.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceForemanOverlay.java:
```diff
@@ -0,0 +1,73 @@
+package net.runelite.client.plugins.blastfurnace;
```
Add license header.

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceConfig.java:
```diff
@@ -52,4 +52,15 @@ default boolean showBarDispenser()
+			description = "If under level 60 smithing, displays time left before requiring payment",
+			position = 3
+	)
+	default boolean showForemanCounter()
```
Instead of having a config for this, display the overlay conditionally if the player is under 60 smithing.

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceForemanOverlay.java:
```diff
@@ -0,0 +1,73 @@
+import java.awt.Graphics2D;
+import java.awt.Dimension;
+
+import static net.runelite.api.MenuAction.RUNELITE_OVERLAY_CONFIG;
```
Same imports issue in this file

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -50,12 +55,17 @@
 {
 	private static final int BAR_DISPENSER = NULL_9092;
 
+	private static final String FOREMAN_PERMISSION_PATTERN = "Okay, you can use the furnace for ten minutes. Remember, you only need half as much coal as with a regular furnace.";
```
Since it's no longer a pattern, use a better var name.

```suggestion
	private static final String FOREMAN_PERMISSION_TEXT = "Okay, you can use the furnace for ten minutes. Remember, you only need half as much coal as with a regular furnace.";
```

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnacePlugin.java:
```diff
@@ -135,4 +154,26 @@ public void onGameStateChanged(GameStateChanged event)
+		{
+			String npcText = Text.sanitizeMultilineText(npcDialog.getText());
+
+			if (npcText.contains(FOREMAN_PERMISSION_PATTERN))
```
Use `.equals()` instead?

bwhite94 replied:
Resolved in 0fadcde

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/blastfurnace/BlastFurnaceForemanOverlay.java:
```diff
@@ -0,0 +1,97 @@
+		long minutes = remainingDuration.toMinutes();
+		long seconds = remainingDuration.getSeconds() % 60;
+
+		return String.format("%01d", minutes) + ":" + String.format("%02d", seconds);
```
I believe this can be `String.format("%01d:%02d", minutes, seconds);`

bwhite94 commented:
Should be in pretty good shape now, let me know if you see anything else that could use some tweaking.

Adam- commented:
I think this would be better as a Timer than an overlay

bwhite94 commented:
> I think this would be better as a Timer than an overlay

Oh wow. I didn't even realize that class existed. Assuming I would add an entry in `GameTimer.java` and then drop the config in `TimersPlugin.java`? Additionally, do you think the Coal bag would be a good image for this? Hover message would be something like "Foreman Fee" or similar for more clarity.

bwhite94 commented:
Refactored to utilize timers per feedback. Removed commits from prior approach to clean up the change history. Much cleaner! Thanks, @Adam- 

The only thing I wasn't sure about a good approach for is `Duration.between(Instant.now(), foremanTimer.getEndTime()).toMinutes() <= 5` which basically just stops checking for the NPC dialog to avoid removing and re-adding the timer until the user closes the dialogue. Something like this was required, to avoid the timer not restarting if paid before completely elapsing.

bwhite94 commented:
Reverted timer plugin changes and extended the timer within the blast furnace plugin per Adam's recommendation in Discord.
